### PR TITLE
Abstracted a CursorResourceManager

### PR DESCRIPTION
**Pull request by rozza**
Part of a wider Command Cursor refactoring

JAVA-5159

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
+/*
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:1]: Done

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java [hunk 2]
```diff
@@ -26,60 +26,51 @@
...
     private final MongoNamespace namespace;
     private final int limit;
     private final Decoder<T> decoder;
     private final long maxTimeMS;
-    private int batchSize;
     @Nullable
     private final BsonValue comment;
-    private CommandCursorResult<T> commandCursorResult;
+
+    private CommandCursorResult<T> commandCursor;
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:68]: `commandCursorResult` seems like a better name, given the type.  Change it back?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java (hunk 2 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java]: Done

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java [hunk 3]
```diff
@@ -256,91 +247,57 @@ public int getMaxWireVersion() {
...
+                                CommandResultDocumentCodec.create(decoder, NEXT_BATCH),
+                                assertNotNull(resourceManager.getConnectionSource()).getOperationContext())));
+                nextServerCursor = commandCursor.getServerCursor();
             } catch (MongoCommandException e) {
                 throw translateCommandException(e, serverCursor);
             }
+
             resourceManager.setServerCursor(nextServerCursor);
-            if (limitReached()) {
+            if (!resourceManager.operable() || limitReached() || nextServerCursor == null) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java]: On further inspection the `nextServerCursor == null` was surplus to requirements as it effectively becomes a noop and releasing of resources is managed by `setServerCursor` above. Removed.

The `!resourceManager.operable()` check fixes a bug where if a cursor is closed in flight (eg State == CLOSE_PENDING) then the cursors are closed. This bug was found when unifying the tests as done in the previous PR.

File: driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java [hunk 4]
```diff
@@ -165,18 +165,20 @@ private void cachePostBatchResumeToken(final AggregateResponseBatchCursor<RawBso
      * @param lastIdConsumer Is {@linkplain Consumer#accept(Object) called} iff {@code rawDocuments} is successfully converted
      *                       and the returned {@link List} is neither {@code null} nor {@linkplain List#isEmpty() empty}.
      */
-    static <T> List<T> convertAndProduceLastId(final List<RawBsonDocument> rawDocuments,
+    static <T> List<T> convertAndProduceLastId(@Nullable final List<RawBsonDocument> rawDocuments,
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java:168]: !resourceManager.operable()

**Comment by rozza** [comment]
Rolled up changes from prior PRs - as was hitting too many rebase issues.

**Comment by rozza** [comment]
Marked as draft - while working through feedback in #399 👍

**Pull request closed without merging**